fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1782,21 +1822,27 @@ private void saveStoreHeader(@Nonnull RecordMetaDataProto.DataStoreInfo storeHea
     }
 
     // Actually (1) writes the index state to the database and (2) updates the cached state with the new state
-    private void updateIndexState(@Nonnull String indexName, byte[] indexKey, @Nonnull IndexState indexState) {
+    private void updateIndexState(@Nonnull Index index, @Nonnull IndexState indexState) {
         // This is generally called by someone who should already have a write lock, but adding them here
         // defensively shouldn't cause problems.
         beginRecordStoreStateWrite();
         try {
             context.setDirtyStoreState(true);
             Transaction tr = context.ensureActive();
             if (IndexState.READABLE.equals(indexState)) {
-                tr.clear(indexKey);
+                tr.clear(indexStateSubspace().pack(index.getSubspaceTupleKey()));
+                if (indexStateIsStoredInOldFormat(recordStoreStateRef.get().getStoreHeader())) {
+                    tr.clear(indexStateSubspaceDeprecated().pack(index.getName()));
```

Comment:
OK, let me do that then. If we go with your way, we could probably also only clear the active key space (suggested earlier).

Reply:
Ok, I gave this a shot. I am reluctant to do it partially, like only here and not in e.g., `addIndexStateReadConflict` or `deleteAllRecords`. How ever several tests end up calling these methods (directly or indirectly) before the record store state is loaded (if even eventually loaded), hence accessing the store header is impossible (leads to NPE). I don't know if that's how they are called in practice too. We cannot just rely on the member `formatVersion` either, due to the window I mentioned in the other comment. (I think we cannot close the window safely, explained in the other comment). Even though, it may be against practice, I'm tempted to update both locations during this upgrade due to the safety. Let me know if you have objections.

Reply:
Which states access these methods before the record store state is loaded? I think that's the error here, and that might also inform whether it's a bug that real users could run into.

There's one other approach here which is to encapsulate creating the index state key into a method, then all of these become `tr.clear(getIndexStateKey())` (or `tr.set`, etc.).

Reply:
Like tests build a record store, using `build`, and the call `deleteAllRecords`. The record store state is not initialized yet. 

> Which states access these methods  before the record store state is loaded? 

I think I don't follow this question.

Reply:
Ah, right, if they are calling `build`, that's the problem. I suspect that the tests are old or wrong, but I could be wrong (they could be trying to test something else that `checkVersion` would get in the way of).

I've had various ideas about what we need to do in this code base to make the whole check version think safer. Ideally, for example, you would only ever get a record store after check version had been called. The `.createOrOpen` family helps a lot with that, but as long as `build` exists, it's a problem. It's possible we just need to outlaw `.build`, though there can occasionally be some reason to do so.

And sorry, I meant "which tests" rather than "which states".

Reply:
Ok, I went with this middle ground. I still want to be in the safe side, but `updateIndexState` only update the key space that it should. So I decided that `upgradeIndexStates` will be the only place that wipes all index states in the old key space. Hence, we will preserve the states in `delateAllRecords`, and we will continue putting read conflicts at both key spaces. Though, `updateIndexState` will only update the correct one based on the store header's format version. Do you buy this?

Reply:
I gave this a try. Several tests fail when I do not put read conflicts on both old and new states. This doesn't happen for the ranges (only when specific indexes do not have read conflicts). Hard to understand where the race is. I'm tempted to move on and keep the read conflicts for both ranges and indexes to remain on the safe side. I'll add a todo and file an issue for it. Let me know if you object. Thanks.

Reply:
Do you have an example of a failing test? It seems to me like we shouldn't be relying on that behavior, so if a test is failing, that could be a sign of a bug.

Reply:
Just repining about a failing test